### PR TITLE
docs: fix broken npm run watch:docs

### LINKED ISSUE #2840: Script "npm run watch:docs" is broken
As discussed in the internal Zulip channel, the script "npm run watch:docs" is broken.

~~~js
npm run watch:docs

> mithril@2.2.2 watch:docs
> node scripts/generate-docs --watch

/www/mithril.js/scripts/_command.js:5
        if (!e.stdout || !e.stderr) throw e
                                    ^

TypeError: Cannot read properties of undefined (reading 'generateSingle')
    at FSWatcher.updateFile (/www/mithril.js/scripts/generate-docs.js:285:24)

Node.js v20.0.0
~~~

### PR DESCRIPTION
**Pull request by tbreuss**
Fixes #2840.

Since the behaviour of the current script was not so obvious, I set the requirements as follows and edited the script therefore.

If the layout.html was deleted:
- throw error "Error: the layout "layout.html" file is required!"

If the layout.html was updated:
- loads the layout file and regenerates all html files in dist

If a navigation file (nav-guides.md|nav-methods.md) was deleted:
- throw error "Error: the navigation file "nav-guides.md" is required!"

If a navigation file was updated:
- loads the navigation files and regenerates all pages in dist

If a markdown file was deleted:
- deletes the html page from dist

If a markdown file was updated:
- regenerates the html page in dist

If a markdown file was added:
- generates the html page in dist

If an asset file was deleted:
- deletes the asset file from dist

If an asset file was updated:
- updates the asset file in dist

If an asset file was added:
- adds the asset file to dist

### CODE REVIEW

File: scripts/generate-docs.js [hunk 1]
```diff
@@ -266,27 +292,45 @@ if (require.main === module) {
...
-			function updateGenerator() {
-				if (timeout == null) return
-				clearTimeout(timeout)
-				genPromise = new Promise((resolve) => {
-					timeout = setTimeout(function() {
-						timeout = null
-						resolve(makeGenerator().then((g) => g.generate()))
-					}, 100)
-				})
+			const generator = (await makeGenerator())
```
Comment (dead-claudia) [inline scripts/generate-docs.js:295]: Minor nit.

```suggestion
			const generator = await makeGenerator()
```

File: scripts/generate-docs.js [hunk 2]
```diff
@@ -266,27 +292,45 @@ if (require.main === module) {
...
+				if (isLayoutFile(relative)) {
+					generator.setLayout(await loadLayoutFile())
+					generator.generate()
+				} else if (isNavigationFile(relative)) {
+					const [guides, methods] = await loadNavigationFiles()
+					generator.setGuides(guides)
+					generator.setMethods(methods)
+					generator.generate()
+				} else {
+					generator.generateSingle(file)
```
Comment (dead-claudia) [inline scripts/generate-docs.js:320]: Every single one of those `.generate()` calls need `await`s as they all return promises.

```suggestion
					generator.setLayout(await loadLayoutFile())
					await generator.generate()
				} else if (isNavigationFile(relative)) {
					const [guides, methods] = await loadNavigationFiles()
					generator.setGuides(guides)
					generator.setMethods(methods)
					await generator.generate()
				} else {
					await generator.generateSingle(file)
```

File: scripts/generate-docs.js [hunk 3]
```diff
@@ -266,27 +292,45 @@ if (require.main === module) {
...
 			async function removeFile(file) {
-				(await genPromise).eachTarget(file, (dest) => fs.unlink(dest))
+				const relative = path.relative(r("docs"), file)
+				if (isLayoutFile(relative)) {
+					throw `Error: the layout file "${relative}" is required!`
+				} else if (isNavigationFile(relative)) {
+					throw `Error: the navigation file "${relative}" is required!`
+				}
+				const relativeDist = relative.replace(/\.md$/, ".html")
+				generator.eachTarget(relativeDist, (dest) => fs.unlink(dest))
```
Comment (dead-claudia) [inline scripts/generate-docs.js:332]: Same as before, different method.

```suggestion
				await generator.eachTarget(relativeDist, (dest) => fs.unlink(dest))
```

File: scripts/generate-docs.js [hunk 4]
```diff
@@ -266,27 +292,45 @@ if (require.main === module) {
...
+			function isLayoutFile(relative) {
+				return relative === "layout.html"
+			}
+
+			function isNavigationFile(relative) {
+				return ["nav-guides.md", "nav-methods.md"].includes(relative)
+			}
+
+			async function updateGenerator() {
+				generator.generate()
```
Comment (dead-claudia) [inline scripts/generate-docs.js:306]: Needs an `await`.

```suggestion
				await generator.generate()
```

**Review by dead-claudia — CHANGES_REQUESTED** [review]
Bit too accidentally parallel.

File: scripts/generate-docs.js (hunk 1 above)
Comment (tbreuss) [reply scripts/generate-docs.js]: fixed

File: scripts/generate-docs.js (hunk 2 above)
Comment (tbreuss) [reply scripts/generate-docs.js]: fixed

File: scripts/generate-docs.js (hunk 4 above)
Comment (tbreuss) [reply scripts/generate-docs.js]: fixed

**Comment by tbreuss** [comment]
@dead-claudia I fixed all code review issues.

**Review by dead-claudia — APPROVED** [review]

**Review by JAForbes — APPROVED** [review]
Looks great @tbreuss

**Comment by dead-claudia** [comment]
Documentation has moved to https://github.com/MithrilJS/docs, and I fixed this there while I assembled that repo.

**Pull request closed without merging**